### ISSUE TITLE
restore: Received an unexpected EOF or 0 bytes from the transport stream

### ISSUE THREAD

**Issue by obeyda** [NuGet/Home#9020]
## Details about Problem

product used (mcr.microsoft.com/dotnet/core/sdk:3.1-buster):

dotnet.exe --version:
```
   docker run --rm mcr.microsoft.com/dotnet/core/sdk:3.1-buster dotnet --version
   3.1.100
```
OS version (i.e. win10 v1607 (14393.321)):
```   
   docker-desktop on windows 10 
   OS Name:                   Microsoft Windows 10 Pro Insider Preview
   OS Version:                10.0.19536 N/A Build 19536

```
## Detailed repro steps so we can see the same problem

restore a project from a solution with a big number of nuget packages and projects (>50 project), no issues for smaller solutions.

## Other suggested things
this works with `--disable-parallel` flag but it takes >20min to complete 

### Logs
[normal verbosity -v:n](https://gist.github.com/obeyda/8eb5e5df3b432ac4537cc3401159a2bc)
[diag verbosity -v:d](https://gist.github.com/obeyda/38e44ce028c200d64854531cfad2495f)

**Comment by dh-ger** [comment]
I can confirm that I am seeing this behavior on the following images:
```
mcr.microsoft.com/dotnet/core/sdk:3.0
mcr.microsoft.com/dotnet/core/sdk:3.0-alpine
```

**Comment by MichaelPetrinolis** [comment]
I also confirm the same behavior, on 3.0:latest and on 3.1:latest on docker for windows 2.2.0.0
--disable-parallel works but it takes a huge amount of time.

**Comment by sergioprates** [comment]
Try add Nuget.config file in project and copy the file to container.

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="NuGet" value="https://api.nuget.org/v3/index.json" />
  </packageSources>
</configuration>
```

**Comment by MichaelPetrinolis** [comment]
my project has a nuget.config file with nuget and a myget source but it makes no difference

**Comment by sergioprates** [comment]
But in dotnet restore, are you passing the file?  

`RUN dotnet restore --configfile ./nuget.config`

**Comment by MichaelPetrinolis** [comment]
it uses it by default. To test it, I added a local package source that does not exist on docker image and the build fails, because it cannot find the source.

**Comment by xVinCe** [comment]
I have the same issue on a single project with ~20 packages referenced (Serilog, EF Core, EPPlus) `mcr.microsoft.com/dotnet/core/sdk:3.1`

I have a slow network at home (~2Mbps), I guess that explains it.
--disable-parallel flag works, takes 6 minutes

**Comment by strizzwald** [comment]
I've got the same issue with 63 packages referenced, using  `mcr.microsoft.com/dotnet/core/sdk:3.1`. Adding `nuget.config` didn't resolve the issue for me.

**Comment by shaunsales** [comment]
Same issue here at 70 package references. I hope this gets some attention, it's slowing down our build & run iteration time in VS quite a bit. 

I've tested on a 10Mbps connection vs 1Gbps connection and it seems bandwidth related - i.e. NuGet tries to connect, times out too quickly and then fails with the error above.

**Comment by obeyda** [comment]
can't we control the max number of concurrent connections? since `--disable-parallel` solves the issue (at least for me, but it takes a long time as i said)

**Comment by aortiz-msft** [comment]
Seems like a dupe of https://github.com/NuGet/Home/issues/8571. Please try the solution discussed in that issue.

**Comment by shaunsales** [comment]
@aortiz-msft I don't think it's related - the error is not the same. I've tested on the same solution with different bandwidth speeds, and this issue occurs only on the slower connection.

**Comment by Gorithow** [comment]
> @aortiz-msft I don't think it's related - the error is not the same. I've tested on the same solution with different bandwidth speeds, and this issue occurs only on the slower connection.

True. And solution with `ulimit -n 8192` did not help me at all.

**Comment by MichaelPetrinolis** [comment]
@aortiz-msft it seems that this is not related to #8571 and should not be closed.

**Comment by obeyda** [comment]
I also tried to increase the maximum number of connection by setting the two values (Linux) :
```
net.ipv4.ip_local_port_range
net.ipv4.tcp_fin_timeout
```

that didn't have any effect.

**Comment by horeeorg** [comment]
I also experienced with this error, but when I tried using tethering from my mobile network, it worked.

**Comment by horeeorg** [comment]
I still experienced with this issue, and I've read, still there is no solution for this issue, could you please open it again?

**Comment by rikkit** [comment]
Also experiencing this error using `mcr.microsoft.com/dotnet/core/sdk:3.1` - `--disable-parallel` fixed it but takes longer. Run inside a Docker build inside WSL2.

**Comment by reazura** [comment]
Im getting this issue locally on `mcr.microsoft.com/dotnet/core/sdk:3.1-buster` as well - which is strange, it was working fine about 2 days ago. I havent tried on another machine/connection yet.

**Comment by SP-SuperPoney** [comment]
I've got the same issue.
Project retrieve Nuget artifacts from Azure at build time.

`RUN dotnet restore "MyProject.csproj"` fail on random packages :

`Determining projects to restore...
Failed to download package 'XXXX' from 'https://pkgs.dev.azure.com/[...]/nuget/v3/XXXX.nupkg'.`

Error can be either :
`Name does not resolve`, `Resource temporarily unavailable`, `The response ended prematurely`. or `Authentication failed because the remote party has closed the transport stream`.

`RUN dotnet restore --disable-parallel "MyProject.csproj"` works but takes a lot of time.

Seem that the workaround with ulimit is per process on linux system.
And The ulimit breaks on the container.
I've added a ulimit just before the dotnet restore and it seems to do the job:

`RUN ulimit -n 8192`
`RUN dotnet restore "MyProject.csproj"`

Don't forget to change on your WSL2 distro the soft file limit: `sudo vi /etc/security/limits.conf` to something like 
`*                soft    nofile          8192`

Notes:
I've a good DSL connexion.
Docker version 19.03.8, build afacb8b.
WSL2 with fresh Debian buster distro.

**Comment by hoopsomuah** [comment]
Why is this issue closed? This is killing my productivity and costing me a lot of money in wasted restores.

**Comment by obeyda** [comment]
@aortiz-msft, is it possible to reopen this? This is not the same issue, the error is not the same, and the proposed solution doesn't work for this problem.
This only happens with slow connections.

**Comment by timba** [comment]
### Confirming the issue

Windows 10 Build 19041.450
Ubuntu 20.04 @ WSL v2
Docker Desktop Engine 19.03.8
Base image mcr.microsoft.com/dotnet/core/sdk:3.1

`--disable-parallel` option works but slows down `dotnet restore` from ~30 sec to 11 min (1 private NuGet feed used additionally to the public NuGet).

The Docker command used:
``` dockerfile
RUN dotnet restore --configfile ./NuGet.config --disable-parallel
```

Without `--disable-parallel` option the command fails with multiple `Received an unexpected EOF or 0 bytes from the transport stream.` errors.

Restore works fine directly on WSL.

**Comment by AndiRudi** [comment]
I have the same error in my docker-compose when I run it locally on a MacBook.

**Comment by kglassmireacr** [comment]
I can vouch that when using private nuget feeds, some of my team members with worse bandwidth often receive this error. 
```
--disable-parallel 
```
works most of the time but even now this is not working.

**Comment by denis-ivanov** [comment]
Any solution?

**Comment by iSeiryu** [comment]
### Confirming the issue
Windows Version: Windows 10 Pro 10.0.19041 Build 19041
WSL 2: Ubuntu 18.04
Docker Desktop Version: 2.3.0.5 (48029)
Docker Engine Version: 19.03.12, build 48a66213fe
Base Image: mcr.microsoft.com/dotnet/core/sdk:3.1-buster

When I run ```dotnet build``` on Windows or inside WSL it works and the whole process takes ~10 seconds. 
But when I run the same command inside Dockerfile via ```docker build . -t test:latest``` it takes 2-5 minutes when it gets to ```Determining projects to restore...``` part and then starts throwing different errors.

#### Errors examples
```
 Failed to download package 'Microsoft.Extensions.Options.ConfigurationExtensions.2.0.0' from 'https://api.nuget.org/v3-flatcontainer/microsoft.extensions.options.configurationextensions/2.0.0/microsoft.extensions.options.configurationextensions.2.0.0.nupkg'.
   Received an unexpected EOF or 0 bytes from the transport stream.
```
```
 Failed to download package 'Microsoft.Extensions.DependencyInjection.Abstractions.3.1.3' from 'https://api.nuget.org/v3-flatcontainer/microsoft.extensions.dependencyinjection.abstractions/3.1.3/microsoft.extensions.dependencyinjection.abstractions.3.1.3.nupkg'.
  The SSL connection could not be established, see inner exception.
     Received an unexpected EOF or 0 bytes from the transport stream.
```
```
Retrying 'FindPackagesByIdAsyncCore' for source 
  The SSL connection could not be established, see inner exception.
    Authentication failed because the remote party has closed the transport stream.
```
```
Name does not resolve
```
```
Failed to download package 'System.Security.Cryptography.Encoding.4.3.0' from 'https://api.nuget.org/v3-flatcontainer/system.security.cryptography.encoding/4.3.0/system.security.cryptography.encoding.4.3.0.nupkg'.
  The download of 'https://api.nuget.org/v3-flatcontainer/system.security.cryptography.encoding/4.3.0/system.security.cryptography.encoding.4.3.0.nupkg' timed out because no data was received for 60000ms.
    Exception of type 'System.TimeoutException' was thrown.
```

Sometimes the whole process works, sometimes it does not at all. The packages it fails on are always different. I only use https://api.nuget.org/

I can even add 
```RUN curl https://api.nuget.org/v3-flatcontainer/microsoft.extensions.options.configurationextensions/2.0.0/microsoft.extensions.options.configurationextensions.2.0.0.nupkg```
inside my Dockerfile and it works instantly.

```--disable-parallel``` option works but takes ~10 minutes.
It also works on the projects that only have a few nuget packages. Although I did not test it much, just ran a few test projects to confirm.

I have a pretty slow internet 4mb/s but this same command works very fast outside of Docker.

**Comment by iSeiryu** [comment]
Linking this issue https://github.com/NuGet/NuGetGallery/issues/5887
It has nothing to do with https://www.myget.org/F
It crashes for pretty much any resource.
Docker is able to pull 100+ Mb images pretty quickly but small nuget packages fail to download.

**Comment by iSeiryu** [comment]
#### Possible workaround
Besides ```--disable-parallel``` the following also works (most of the time).
Inside Dockerfile copy only the main project(s) first
```dockerfile
COPY ["YourProject.csproj", "ProjectDir/"]
RUN dotnet restore "ProjectDir/YourProject.csproj"

# you can also add a nuget.config manually
RUN dotnet restore "ProjectDir/YourProject.csproj" --configfile nuget.config

# and then copy the rest of the files do the build
COPY . .
RUN dotnet build "ProjectDir/YourProject.csproj" -c Release
```

It outputs a lot of
```
Determining projects to restore...
Skipping project "/src/SomeProject/SomeProject.csproj" because it was not found.
```
But when it gets to dotnet build step it just works 
```
Determining projects to restore...
Restored /src/SomeProject/SomeProject.csproj (in 5.38 sec).
```

There is also this article https://andrewlock.net/optimising-asp-net-core-apps-in-docker-avoiding-manually-copying-csproj-files/

**Comment by iSeiryu** [comment]
mcr.microsoft.com/dotnet/sdk:5.0
30 MB/s channel

Still the same set of errors even on small projects. The workarounds above work but are slow.
`curl`ing the same packages inside the same Dockerfile happens instantly.

**Comment by DOMZE** [comment]
Same problem here. with the following:
Docker version 20.10.5, build 55c4c88
WSL2: Ubuntu (Default)
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.1 LTS
Release:        20.04
Codename:       focal

running
`RUN ulimit -n 8192` did not fix the problem
Only thing that worked is using `--disable-parallel`

using docker `mcr.microsoft.com/dotnet/sdk:5.0`

**Comment by erdembayar** [comment]
> Same problem here. with the following:
> Docker version 20.10.5, build 55c4c88
> WSL2: Ubuntu (Default)
> Distributor ID: Ubuntu
> Description: Ubuntu 20.04.1 LTS
> Release: 20.04
> Codename: focal
> 
> running
> `RUN ulimit -n 8192` did not fix the problem
> Only thing that worked is using `--disable-parallel`
> 
> using docker `mcr.microsoft.com/dotnet/sdk:5.0`

Please upvote by pressing "Like" on issue to prioritize this issue, we'll fix in coming sprints.
Meantime please disable parallel connection work around. `mcr.microsoft.com/dotnet/core/sdk:3.1 - --disable-parallel`

**Comment by doniyorniazov** [comment]
I have the same problem. I have a ton of packages to restore. With using --disable-parallel I need to wait at least 30 - 45 mins. Is there any other solutions? Run ulimit -n 8192 doesn't f x the problem.   Sometimes it works fine even without --disable-parallel. I'm not sure why???

**Comment by doniyorniazov** [comment]
Hi All, I had also similar problem. It seems its related to internet speed also. **But I tried to remove all my images and containers and and also logout and log in back. I built a new image and it worked again just like before**. Let me know if that will help to any of you.

UPDATE: 
It came back again. I'm being able to reproduce it every time when I want to build my image.

**Comment by zivkan** [comment]
6 months ago I had a brief chat with someone who maintains Microsoft's .NET docker images (at https://github.com/dotnet/dotnet-docker).

The recommendation I got at the time was to search and/or post in https://github.com/dotnet/dotnet-docker/discussions. Particularly if you find that restore works on the host machine, but not in docker. Docker has network configuration that can cause it to behave differently to the host machine, which the NuGet team are not experts on.

If you can reproduce the problem on the host machine, then here (NuGet) is the best place to report the issue.

**Comment by pmozbert** [comment]
This is still a problem  - I'm getting the unexpected EOF error.
Adding --disable-parallel fixed it for me.

**Comment by kartheekp-ms** [comment]
Reading from the comments from the customers I am assuming it may be related to the default concurrency setting. Users have mentioned that disable parallelism fixes the issue. We set default degree of concurrency to `16` https://github.com/NuGet/NuGet.Client/blob/e3016693e01536996921d613152648d66eb0ce99/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRequest.cs#L20. I am assuming if this setting should be [Environment.ProcessorCount​](https://docs.microsoft.com/en-us/dotnet/api/system.environment.processorcount?view=net-6.0). I think xUnit does the same thing when we enable [parallelism for tests](https://github.com/xunit/xunit/blob/56bd72f9b99e1eb9871fa739e0c438029154c28a/src/xunit.v3.core/Sdk/v3/Runners/XunitTestAssemblyRunner.cs#L248). If this makes sense, we can ask customers to execute restore with [​`maxHttpRequestsPerSource`​](https://github.com/NuGet/Home/wiki/Config-Max-Http-Request-In-NuGet#solution) set to ​​`processor count​​` and see if they can reproduce the failure again. I learnt from this [wiki](https://github.com/NuGet/Home/wiki/Config-Max-Http-Request-In-NuGet#current-code) that ​`Package.config use maxDegreeOfParallelism which is set to Environment.ProcessorCount​`.

``` xml
<?xml version='1.0'?>
<configuration>
    <config>
        <add key='maxHttpRequestsPerSource' value='x' />
    </config>
</configuration>
```
I have not tried that myself, another option is [​`dotnet trace​​`](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-trace) command can be executed to capture events from`Microsoft-System-Net-Http`​ event provider ​during restore. I learnt about this command from https://developers.redhat.com/blog/2019/12/23/tracing-net-core-applications#dotnet-trace doc.

cc @zivkan in case it helps.

**Comment by zivkan** [comment]
@kartheekp-ms don't confuse CPU bound work with network IO bound work.  CPU bound work it makes sense to limit to processor count, because having more threads (that are not blocked) than CPU cores will mean the operating system will keep context switching threads, and the overhead of doing that will decrease application throughput compared to limiting therads to processor count (and using good async code).  Also, the default concurrency you linked to in `RestoreRequest` appears to be the number of projects restored in parallel, which is not exactly the same as max requests per HTTP source. Depending on the number of packages, and the "shape" of the graph, it influences greatly the number of max parallel HTTP requests NuGet might attempt.  Nuget doesn't create threads manually, it just uses the ThreadPool, which itself manages thread counts, trying to keep only one thread per CPU core, but will increase when it detects threads blocked on synchronous APIs.

On the other hand, network IO should be unrelated to processor count. You don't want a low CPU count machine to underutilize its network, and you don't want a high CPU count machine to be a bad netizen and swamp a server with too many parallel requests.

Other things to note, [NuGet sets `ServicePointManager.DefaultConnectionLimit` to 64 on .NET Framework](https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/src/NuGet.Core/NuGet.Common/NetworkProtocolUtility.cs#L13). ServicePointManager exists on .NET (Core), and I thought it was a no-op, but looking though dotnet/runtime, it looks like it might still be used by the non-sockets based http handler, but it's not used by the socket's based http handler, which has been default since .NET Core 3.1 I think (maybe .NET Core 2). The way to set concurrency limits on HttpClient in .NET Core is to use the properties on the HttpClientHandler, which NuGet doesn't set, hence it uses the runtime default.

I'm not 100% sure about .NET Core, but the expected behaviour when your app tries to make more parallel requests that the concurrency limit on the HttpClient, is that the HttpClient will just wait until other requests finish.

All this just to say that I disagree that max concurrency being too high from a client runtime will result in "Recieved an unexpected EOF or 0 bytes from the transport stream". I feel like something else is going on.

It's possible that the server limits the number of parallel sockets open per IP address, and force closes them if the limit is exceeded, but in that case I'd expect the same problem to occur when restoring on the host machine, not just in docker containers (if both the host machine and docker contain have sockets open at the same time, it might just be a coincidence that it's noticed on the docker container only, and not the host).

So, in my opinion, it's complicated and we don't have anywhere near enough information. Changing the `maxHttpRequestsPerSource` is a great idea, just keep in mind if you have multiple sources on the same host (for example multiple Azure DevOps feeds), the number of parallel HTTP sockets open to Azure Artifacts will be proportional to the number of sources, not just the setting in nuget.config.

The last time I attempted to investigate, I couldn't replicate the problem, so at this time I feel like we need a highly technical customer who is impacted and is willing to help us with some trial-and-error experimentation, plus trying things like `dotnet trace` to gather more information.

**Comment by hmz777** [comment]
I was working with [eShopOnContainers](https://github.com/dotnet-architecture/eShopOnContainers) and what solved it for me, was to add this to `NuGet.config` (5 is a random number you can experiment with more but 5 did the trick and It didn't take too much time).
```
<configuration>
    <config>
        <add key='maxHttpRequestsPerSource' value='5' />
    </config>
</configuration>
```
Then `docker-compose build`.

**Comment by JesseRoxit** [comment]
Not sure what the exact cause is but, for me, after uninstalling Fiddler it behaves as expected again. So people who still face this issue and have Fiddler installed, you can try uninstalling it.

**Comment by erdembayar** [comment]
@nkolev92 @aortiz-msft 
Unassigned me for load balancing.

**Comment by jeffkl** [comment]
I've spent a lot of time trying to reproduce this issue and https://github.com/NuGet/Home/issues/10275#issuecomment-1185759724 but haven't yet had it happen.  Does anyone have precise repro steps I can follow to investigate this further?

**Comment by aortiz-msft** [comment]
We haven't been able to reproduce the issue, and we haven't received more information in the last year. Please open a new issue if you see this problem again.
